python/ray/tune/ray_trial_executor.py

```diff
@@ -599,19 +599,21 @@ def restore(self, trial, checkpoint=None):
                 "Trial {}: Unable to restore - no runner found.".format(trial))
         value = checkpoint.value
         if checkpoint.storage == Checkpoint.MEMORY:
-            assert not isinstance(value, Checkpoint), type(value)
+            # Note that we don't store the remote since in-memory checkpoints
+            # don't guarantee fault tolerance and don't need to be waited on.
             trial.runner.restore_from_object.remote(value)
         else:
             logger.info("Trial %s: Attempting restore from %s", trial, value)
-            with warn_if_slow("get_current_ip"):
-                worker_ip = ray.get(trial.runner.current_ip.remote(),
-                                    DEFAULT_GET_TIMEOUT)
-            with warn_if_slow("sync_to_new_location"):
-                trial.sync_logger_to_new_location(worker_ip)
-            with warn_if_slow("restore_from_disk"):
-                # TODO(ujvl): Take blocking restores out of the control loop.
-                ray.get(trial.runner.restore.remote(value))
-        trial.last_result = checkpoint.result
+            if trial.upload_dir is None and trial.sync_on_checkpoint:
+                # This provides FT backwards compatibility in the
+                # case where an upload directory is not provided.
+                logger.warning("Trial %s: Reading checkpoint into memory.")
+                data_dict = TrainableUtil.pickle_checkpoint(value)
+                remote = trial.runner.restore_from_object.remote(data_dict)
+            else:
+                remote = trial.runner.restore.remote(value)
+            self._running[remote] = trial
+            trial.on_begin_restore(checkpoint)
```

Comment:
it's not immediately clear to me what this `begin_restore` is doing (obviously I can read it but does it need to be abstracted?

Reply:
oh are you saying i should just do `trial.restoring_from = checkpoint`? sure that makes sense, initially i had a more complex implementation which is why it was abstracted away.